Accenture/reactive-interaction-gateway
# NATS support

Closes #297 

## Description

Support for [NATS](https://nats.io). This includes:

- forwarding events from NATS topics to RIG clients, and
- converting HTTP requests received from clients into NATS messages,
  - without waiting for the response,
  - with waiting for the response.

A notable difference to the Kafka and Kinesis implementations is that in the proxy case the response it received via the NATS request/response mechanism and _not_ using the correlation ID. Thought it's good to keep the correlation ID in there anyway to aid debugging; downside is that this may be confusing for users. Due to this semantic differences I've deliberately _not_ documented this feature right now and there is also no integration test for it yet. If need arises (that is, someone creates a GitHub issue for this) we can revisit.

## Open issues


## How to test this

Run the integration test with `./integration_tests/nats_tests/run.sh`. Read on for manual testing..

### Manual testing

Start RIG:

```bash
NATS_SERVERS=localhost:4222 NATS_SOURCE_TOPICS=rig1 iex -S mix
```

RIG should tell you that it can't connect. Now start a NATS server:

```bash
docker run -d --rm -p 4222:4222 nats
```

RIG should now stop yelling around. Connect to RIG and subscribe to the events of type `test`:

```bash
http --stream :4000/_rig/v1/connection/sse\?subscriptions='[{"eventType":"test"}]'
```

Now publish a matching CloudEvent - you could use a NATS client for this but instead we'll use a second instance of RIG:

```bash
export PROXY_CONFIG_FILE=$(cat <<EOF
[
  {
    "id": "my-service",
    "version_data": {
      "default": {
        "endpoints": [
          {
            "id": "my-endpoint",
            "method": "POST",
            "path": "/",
            "target": "nats",
            "topic": "rig1"
          }
        ]
      }
    },
    "proxy": {
      "use_env": true,
      "target_url": "API_HOST",
      "port": 3000
    }
  }
]
EOF
)

# The other RIG instance is still running, so we can't use the standard ports:
export INBOUND_PORT=4100 INBOUND_HTTPS_PORT=4111 API_HTTP_PORT=4110 API_HTTPS_PORT=4101
# By default, RIG consumes from the "rig" topic. We want the other instance to consume 
NATS_SERVERS=localhost:4222 NATS_SOURCE_TOPICS=rig2 iex -S mix
```

Then in another terminal:

```bash
cat <<EOF | http post :4100 
{
  "specversion": "0.2",
  "type": "test",
  "source": "test",
  "id": "069711bf-3946-4661-984f-c667657b8d85",
  "time": "2020-04-05T17:31:00Z",
  "contenttype": "application/json",
  "data": {
    "foo": "bar"
  }
}
EOF
```

You should see this now in the SSE terminal.
